design/sketch.webidl

```diff
@@ -0,0 +1,532 @@
+};
+
+// InputState
+typedef u32 WebGPUIndexFormatEnum;
+interface WebGPUIndexFormat {
+    const u32 UINT16 = 0;
+    const u32 UINT32 = 1;
+};
+
+typedef u32 WebGPUVertexFormatEnum;
+interface WebGPUVertexFormat {
+    const u32 FLOAT_R32_G32_B32_A32 = 0;
+    const u32 FLOAT_R32_G32_B32 = 1;
+    const u32 FLOAT_R32_G32 = 2;
+    const u32 FLOAT_R32 = 3;
+    // TODO other vertex formats
+};
+
+typedef u32 WebGPUInputStepModeEnum;
+interface WebGPUInputStepMode {
+    const u32 VERTEX = 0;
+    const u32 INSTANCE = 1;
+};
+
+dictionary WebGPUVertexAttributeDescriptor {
+    u32 shaderLocation;
+    u32 inputSlot;
+    u32 offset;
+    WebGPUVertexFormatEnum format;
+};
+
+dictionary WebGPUVertexInputDescriptor {
+    u32 inputSlot;
+    u32 stride;
+    WebGPUInputStepModeEnum stepMode;
+};
+
+dictionary WebGPUInputStateDescriptor {
+    WebGPUIndexFormatEnum indexFormat;
+
+    sequence<WebGPUVertexAttributeDescriptor> attributes;
+    sequence<WebGPUVertexInputDescriptor> inputs;
+};
+
+interface WebGPUInputState {
+};
+
+// ShaderModule
+dictionary WebGPUShaderModuleDescriptor {
+    ArrayBuffer code;
+};
+
+interface WebGPUShaderModule {
+};
+
+// AttachmentState
+dictionary WebGPUAttachmentStateDescriptor {
+    sequence<WebGPUTextureFormatEnum> formats;
+    // TODO other stuff like sample count etc.
+};
+
+interface WebGPUAttachmentState {
+};
+
+// Common stuff for ComputePipeline and RenderPipeline
+typedef u32 WebGPUShaderStageEnum;
+interface WebGPUShaderStage {
+    const u32 VERTEX = 0;
+    const u32 FRAGMENT = 1;
+    const u32 COMPUTE = 2;
+};
+
+dictionary WebGPUPipelineStageDescriptor {
+    WebGPUShaderModule module;
+    WebGPUShaderStageEnum stage;
+    DOMString entryPoint;
+    // TODO other stuff like specialization constants?
+};
+
+dictionary WebGPUPipelineDescriptorBase {
+    WebGPUPipelineLayout layout;
+    sequence<WebGPUPipelineStageDescriptor> stages;
+};
+
+// ComputePipeline
+dictionary WebGPUComputePipelineDescriptor : WebGPUPipelineDescriptorBase{
+};
+
+interface WebGPUComputePipeline {
+};
+
+// WebGPURenderPipeline
+typedef u32 WebGPUPrimitiveTopologyEnum;
+interface WebGPUPrimitiveTopology {
+    const u32 POINT_LIST = 0;
+    const u32 LINE_LIST = 1;
+    const u32 LINE_STRIP = 2;
+    const u32 TRIANGLE_LIST = 3;
+    const u32 TRIANGLE_STRIP = 4;
+};
+
+dictionary WebGPURenderPipelineDescriptor : WebGPUPipelineDescriptorBase{
+    WebGPUPrimitiveTopologyEnum primitiveTopology;
+    sequence<WebGPUBlendState> blendState;
+    WebGPUDepthStencilState depthStencilState;
+    WebGPUInputState inputState;
+    WebGPUAttachmentState attachmentState;
+    // TODO other properties
+};
+
+interface WebGPURenderPipeline {
+};
+// ****************************************************************************
+// COMMAND RECORDING (Command buffer and all relevant structures)
+// ****************************************************************************
+
+typedef u32 WebGPULoadOpEnum;
+interface WebGPULoadOp {
```

Comment:
`DONT_CARE`?

Reply:
I'd lean for a defer: we don't want to introduce uninintialized reads in WebGPU.

Reply:
This is not an obvious question to dismiss. Given all memory initialized on allocation (conceptually) by WebGPU, it's not a security concern to have the contents from previous rendering showing up, it's a portability concern. And the cost of that is not being able to optimize the case where user knows they are rendering a full-screen quad (for example) so they don't want either clearing or storing the contents.

Reply:
The punctuation made this unclear. I was suggesting we defer the discussion and also gave our PoV.

Reply:
> I'd lean for a defer: we don't want to introduce uninintialized reads in WebGPU.

I agree. However, a small point. Can an application directly read from current draw attachment? That would introduce a feedback loop, and we probably would want to eliminate such situations anyway. There's a possibility to read that data "indirectly", e.g. via blending, but we can validate against that too (I can't imagine a well behaved app that would try to draw with blending into an attachment with `DONT_CARE` load op).